### ISSUE TITLE
Uncaught TypeError in setCanvasElementSizeCallingThread

### ISSUE THREAD

**Issue by floppyhammer** [emscripten-core/emscripten#20246]
The error says

`Uncaught TypeError: Cannot read properties of null (reading '0')`

The error comes from the code below.

```js
var autoResizeViewport = false;
if (canvas.GLctxObject && canvas.GLctxObject.GLctx) {
  var prevViewport = canvas.GLctxObject.GLctx.getParameter(2978);
  autoResizeViewport =
    prevViewport[0] === 0 &&
    prevViewport[1] === 0 &&
    prevViewport[2] === canvas.width &&
    prevViewport[3] === canvas.height;
}
```

It seems prevViewport is 0.

I'm not really familiar with the js side. Is the js file that contains `setCanvasElementSizeCallingThread` generated by emsdk? I'm not sure if this is an emsdk issue or a user issue.

**Comment by sbc100** [comment]
Yes the code for `setCanvasElementSizeCallingThread` comes from the emscripten library: https://github.com/emscripten-core/emscripten/blob/c6a1b76d0311212a1bd5254a182e80f8d9fe4e87/src/library_html5.js#L2317

Can you give some more information about what you are building?  What you are the link flags you using?  What is the callstack when you see this error?   Are you building with assertions enabled?   (either `-O0` or `-sASSERTIONS`).

**Comment by floppyhammer** [comment]
1. I'm building some sort of video editing library. It has a quite big codebase and previously runs in native environments. It has been compiled into wasm recently. The app is a video editor running in browsers.

2. Linker flags:

```
-pthread -lembind -s MAX_WEBGL_VERSION=2 -s FULL_ES3=1 -s INITIAL_MEMORY=128MB -s MAXIMUM_MEMORY=2GB -s ALLOW_MEMORY_GROWTH=1 -s USE_PTHREADS=1 -s OFFSCREENCANVAS_SUPPORT=1 -s OFFSCREEN_FRAMEBUFFER=1 -sPTHREAD_POOL_SIZE=30 -s WASM=1 -lwebsocket.js -sPROXY_POSIX_SOCKETS -s MODULARIZE=1 -s EXPORT_ES6=1 -s ENVIRONMENT='web,worker' -s EXPORTED_RUNTIME_METHODS=FS -s FORCE_FILESYSTEM -s WASM_BIGINT -lidbfs.js -s TEXTDECODER=0
```

3. Unfortunately there's no detailed callstack. I got most info from this screenshot.

![image](https://github.com/emscripten-core/emscripten/assets/28705694/5a8a2cd2-e3c6-407e-92e2-f837d849f649)

4. Assertions is disabled and -O2 is enabled.

**Comment by sbc100** [comment]
Can you build with without `-O2` (i.e. without optimizations and with assertions enabled).  It may provide useful information.

As an asside, regarding your settings:

- you don't need `-s USE_PTHREADS` (since you have `-pthread` in there
- you don't need `-s WASM` since that is the default
- `-sPTHREAD_POOL_SIZE=30` is likely going to use a lot of resources.  Do you need to start 30 threads before your program starts up?
